### DISCUSSION TITLE
Building GDL on M1 Mac using 'way1' fails with plplot

### DISCUSSION THREAD

**Discussion by DJCWinyard** [gnudatalanguage/gdl#1392] [Q&A]
Hi - I decided to try to build GDL on an ARM64 Mac. I followed the instructions in 'way1' here: https://github.com/gnudatalanguage/gdl/wiki/GDL-on-OSX

I have up-to-date XCode (14.1) with the cli-tools installed. Homebrew installed fine. The build_gdl.sh `prep` and `configure` modes seemed to work correctly and plplot is installed into homebrew by these scripts. (`plplot5.15.0 -> ../Cellar/plplot/5.15.0_4/lib/plplot5.15.0`).

But when I run `build_gdl.sh build` - there are linker errors, saying plplot cannot be found:

> [ 18%] Built target wxwidgets
> Consolidate compiler generated dependencies of target gdl
> ld: library not found for -lplplot
> ld: library not found for -lplplot
> ld: library not found for -lplplot
> ld: library not found for -lplplot
> clang-15clang-15clang-15: : : clang-15: error: error: error: error: linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)
> 
> make[2]: make[2]: *** [src/plplotdriver/xwin.so] Error 1*** [src/plplotdriver/mem.so] Error 1
> 
> make[2]: make[2]: *** [src/plplotdriver/svg.so] Error 1*** [src/plplotdriver/ps.so] Error 1
> 
> make[1]: *** [src/plplotdriver/CMakeFiles/mem.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make[1]: *** [src/plplotdriver/CMakeFiles/ps.dir/all] Error 2
> make[1]: *** [src/plplotdriver/CMakeFiles/svg.dir/all] Error 2
> make[1]: *** [src/plplotdriver/CMakeFiles/xwin.dir/all] Error 2`
> 

I am using bash for the build as advised. 
My clang version is 14.0 - (from `clang --version`). The OS is MacOS Ventura 13.01

There seems to definitely be a problem with plplot because if I re-run `build_gdl.sh prep` it reports that everything is installed properly, apart from plplot which it reinstalls. It always reinstalls plplot every time I run it (regardless of whether I tried to run the `build` script or not). It reports no errors building plplot, but always reinstalls it.:

...

> Warning: qhull 2020.2 is already installed and up-to-date.
> To reinstall 2020.2, run:
>   brew reinstall qhull
> [build_gdl.sh] Installing plplot...
> /Users/doktor/Library/Caches/Homebrew/downloads/8bcb5e15a35f7e8c535ffbdf4990fae37fec493cd9a11ec5759c0bac9118b8e8--plplot--5.15.0_4.arm64_monterey.bottle.tar.gz
> ==> Downloading https://downloads.sourceforge.net/project/plplot/plplot/5.15.0%20Source/plplot-5.15.0.tar.gz
> Already downloaded: /Users/doktor/Library/Caches/Homebrew/downloads/cf8af0e0c2fc17518b8c64a2b20f4ebef5f3dcdf989f0be94e062b4aa19ca689--plplot-5.15.0.tar.gz
> ==> Reinstalling plplot 
> ==> cmake .. -DCMAKE_INSTALL_LIBDIR=/opt/homebrew/Cellar/plplot/5.15.0_4/lib -DPL_HAVE_QHULL=OFF -DENABLE_ada=OFF -DENABLE_d=OFF -DENABLE
> ==> make
> ==> make install
> 🍺  /opt/homebrew/Cellar/plplot/5.15.0_4: 316 files, 8.8MB, built in 18 seconds

**Comment by alaingdl** [comment]
OK, thanks you for this very rich report. Except from (re)compiled by yourself plplot then GDL, I see no way to help you now.

On my side, I have to say that yesterday, on "old" x86 Mac with OS 12.3.1 I did succeed to compile with the script. Now
Cmake is in version 3.24.3. When Cmake was in version 3.23 I was not able to compile on the same machine !

**Reply by DJCWinyard** [reply]
Hi Alain - thank you for the quick response! So - you think for some reason the plplot library is not being created properly by `brew install plplot`? I checked my Cmake version and it is 3.24.3. Unfortunately I am not very familiar with Cmake and linkers - can you suggest what steps/commands I need to issue to finish the build by hand?

In the meantime I am trying it on an Ubuntu VM in Parallels also...

**Comment by alaingdl** [comment]
context : working remotely on a OSX machine, ssh -X, xclock OK)
and Apple clang version 13.1.6 (clang-1316.0.21.2)

0/ compilation with the script OK (but see 3/ below)
1/ with the current git version, when using cmake without the script, I have the same issue than you (including providing the path to a plplot I compiled by my-self) (complaining about linking)
2/ going back to 1.0.0rc3 and compiling with
 cmake .. -DREADLINE=off -DWXWIDGETS=off -DOPENMP=OFF -DHDF=OFF
(then using plplot provided by Brew)
I have a working GDL and I can display (X) windows through network :smiley: 

Then : 
-- plplot provided by Brew is not bad, at least for X11 
-- I suspect CMake files for plplot are not perfect
-- small old rampant bug in readline detection in OSX  
-- unclear for me how local plplot drivers in current GDL git are supposed to work and to be link

3/ last information : we are not far for a great step forward : 
-- as mention in 0/ above, script OK here with the git version (then with wxwidgets)
-- but when launching it : 
```
GDL> help,/dev
Available Graphics Devices: NULL PS SVG Z X 
Current graphics device: X
GDL> plot, findgen(10)
./src/common/dpycmn.cpp(130): assert ""IsOk()"" failed in GetGeometry(): invalid wxDisplay object
Collecting stack trace information, please wait...Trace/BPT trap: 5
```
I suspect just a _detail_ is missing

Then if you want to have a working version on you OSX + M1 you can get the 1.0.1rc3. We did that last year with my student. I can also mention few days ago with a colleague we did install a working version using MacPort. (but quite sure is was not on a M1 machine)

**Comment by GillesDuvert** [comment]
Hi,
The error message just says that the plplot library is not found at link time.
You did not mention it, but doing a `brew upgrade` may help on such a new architecture.
if you have determined that libplplotcxx.15.0.0.dylib etc are in /usr/local/lib/ then check the value of LIBRARY_PATH , perhaps setting it to the correct value will help:
`export LIBRARY_PATH=/usr/local/lib/`
Another way may be to define `LDFLAGS` to `-L/usr/local/opt/plplot/lib`
HIH

**Comment by GillesDuvert** [comment]
My previous comment was about the link problem only.
I see no advantage to use an obsolete (1.0.0rc3) version instead of the development one because the build_gdl.sh of 1.0.0rc3 is too old with respect of the current versions of libraries distributed by brew. Besides, the new build_gdl.sh will  'brew reinstall plplot' adding the necessary flags to get the correct drivers in the plplot library, not the one provided by brew that cannot work with the current version of GDL.
But I may be mistaken, there are too many dependencies and changes in plplot, homebrew, gdl and osx that anything can happen.
I believe that, after a failing compile or install with the last version of build_gdl.sh, or if the last gdl does not work properly, you can change only the gdl sources (going back to 1.0.0rc3 for example) and just do a `build_gdl.sh compile` and  build_gdl.sh install` again.

For macOSX, it would be simpler to provide an installer as we do for Windows see https://github.com/gnudatalanguage/gdl/releases ... this is still pending.
